src/logic/safe/transactions/txMonitor.ts

```diff
@@ -0,0 +1,88 @@
+import { TransactionReceipt } from 'web3-core'
+
+import { web3ReadOnly } from 'src/logic/wallets/getWeb3'
+import { sameAddress } from 'src/logic/wallets/ethAddresses'
+import { sameString } from 'src/utils/strings'
+
+type TxMonitorProps = {
+  sender: string
+  hash: string
+  data: string
+  nonce?: number
+  gasPrice?: string
+}
+
+type TxMonitorOptions = {
+  delay?: number
+}
+
+/**
+ * Recursively inspects a pending tx. Until it's found, and returns the mined tx receipt
+ *
+ * @param {object} txParams
+ * @param {string} txParams.sender
+ * @param {string} txParams.hash
+ * @param {string} txParams.data
+ * @param {number | undefined} txParams.nonce
+ * @param {string | undefined} txParams.gasPrice
+ * @param {function(txReceipt: TransactionReceipt): void} cb - called with the tx receipt as argument when tx is mined
+ * @param {object} options
+ * @param {number} options.delay
+ */
+export const txMonitor = async (
+  { sender, hash, data, nonce, gasPrice }: TxMonitorProps,
+  cb: (txReceipt: TransactionReceipt) => void,
+  options?: TxMonitorOptions,
+): Promise<void> => {
+  setTimeout(async () => {
+    if (nonce === undefined || gasPrice === undefined) {
+      // this block is accessed only the first time, to lookup the tx nonce and gasPrice
+      // find the nonce for the current tx
+      const transaction = await web3ReadOnly.eth.getTransaction(hash)
+
+      if (transaction !== null) {
+        // transaction found
+        return txMonitor({ sender, hash, data, nonce: transaction.nonce, gasPrice: transaction.gasPrice }, cb, options)
+      } else {
+        return txMonitor({ sender, hash, data }, cb, options)
```

Comment:
It looks like `txMonitor` itself also implements a recursive timeout loop.
So we could use `waitWithTimeout` here as well, and use a promise instead of the callback:

```
export const txMonitor = (txMonitorProps) => {
  return new Promise((resolve, reject) => {
    waitWithTimeout(() => { ... }, resolve, reject)
  })
}
```

Reply:
I solved it in a simpler way, you can see the change in the last commit

Reply:
That was smart! 👏